### PR TITLE
Functions: tracking subscriptions

### PR DESCRIPTION
**Pull request by pinebit**

### CODE REVIEW

File: core/scripts/functions/templates/oracle.toml
Comment (KuphJr) [inline core/scripts/functions/templates/oracle.toml:48]: We have multiple places in the jobspec where there are duplicate definitions of the router contract address.  How hard would it be to exclusively use the router address specified under `contractID` everywhere?
(This is out of scope for this PR, but may be a good backlog ticket as it may add misconfiguration risk)

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (KuphJr) [inline core/services/gateway/handlers/functions/subscriptions.go:141]: The current contract implementation reverts if there are zero subscriptions. https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol#L269

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (KuphJr) [inline core/services/gateway/handlers/functions/subscriptions.go:185]: When would this case happen? Should we log an error here?

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (KuphJr) [inline core/services/gateway/handlers/functions/subscriptions.go:187]: What is the purpose of this line?

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (KuphJr) [inline core/services/gateway/handlers/functions/subscriptions.go:188]: Maybe we should check the balance and only save the subscription with the highest balance here? Although maybe we should also re-work this as we are currently holding the rwMutext here.

File: core/scripts/functions/templates/oracle.toml
Comment (pinebit) [reply core/scripts/functions/templates/oracle.toml:48]: Agreed, this would be a good improvement.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/subscriptions.go:44]: Good point, İ need to improve this

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/subscriptions.go:185]: If subscription is cancelled I guess. I need to test it better though

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/subscriptions.go:187]: Golang linter error: you cannot take reference for a loop variable unless your created a locally scoped copy like I did.

File: core/scripts/functions/templates/oracle.toml
Comment (bolekk) [reply core/scripts/functions/templates/oracle.toml:48]: Add a ticket to tech debt epic please!

File: core/services/functions/connector_handler.go
Comment (bolekk) [inline core/services/functions/connector_handler.go:73]: This needs to check against sub balance. So we need required balance in the config and probably a more specific API than GetSubscription().

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:24]: To follow allowlist config fields, I'd suggest "UpdateFrequencySec", "UpdateTimeoutSec" and "UpdateRangeSec".

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:22]: Allowlist has "ContractAddress" but "Router" is slightly better. So up to you if you prefer consistency or a slightly more specific name.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:27]: Missing MinimumBalance field. It could go here or maybe better in the top-level FunctionsHandlerConfig as the subscription fetcher is technically independent of balance checking logic. If you were planning to do it in a separate PR that's fine too.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [reply core/services/gateway/handlers/functions/subscriptions.go:44]: We should store all of them.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:137]: We could cache this and use previous value if fetching fails.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:161]: Call it once before the for loop so node doesn't need to always wait for QueryFrequencySec on startup.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [reply core/services/gateway/handlers/functions/subscriptions.go:185]: Yes, this is for cancelled subs. But we shouldn't continue, we need to still store it as cancelled. Otherwise we'll keep using previous values.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [reply core/services/gateway/handlers/functions/subscriptions.go:188]: Let's save all but have an API to retrieve highest balance. In the future we are also going to look at flags stored within subscriptions.

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/functions/connector_handler.go
Comment (bolekk) [inline core/services/functions/connector_handler.go:39]: Planning to also add to the Gateway handler?

Comment (pinebit) [reply core/services/functions/connector_handler.go:39]: Do you think it needs to be enabled in GW handler too?

Comment (bolekk) [reply core/services/functions/connector_handler.go:39]: I don't think it's a must. But it should be very easy to add there as optional and the more bad requests we can stop in Gateway, the less unnecessary load we put on the nodes.

File: core/scripts/functions/templates/oracle.toml
Comment (pinebit) [reply core/scripts/functions/templates/oracle.toml:48]: FUN-921

**Review by KuphJr — DISMISSED** [review]
One missed log line.

Other that that, my concerns have been addressed, so it LGTM.  Bolek should review & make sure his comments have been addressed before we merge.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:197]: Add log messages with number of subs updated.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/subscriptions.go:138]: Log updated count.

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/subscriptions.go:156]: We have cached value. My idea is that every time the cycle reaches the end, we update count again to see if there are new subs.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/subscriptions.go:198]: This will always be equal to `UpdateRangeSize` except for the tail. Not sure this is useful at all.

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:76]: Isn't this an opportunity to DDoS?

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go]: Done

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [reply core/services/gateway/handlers/functions/subscriptions.go:198]: you're right

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:81]: Done for Gateway.

File: core/services/functions/connector_handler.go
Comment (bolekk) [inline core/services/functions/connector_handler.go:45]: It's not Wei but Juels.

File: core/services/gateway/handlers/functions/handler.functions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/handler.functions.go:45]: Juels

File: core/services/gateway/handlers/functions/handler.functions.go
Comment (bolekk) [inline core/services/gateway/handlers/functions/handler.functions.go:104]: Maybe better to have Juels in config? I'm worried about precision.

File: core/services/gateway/handlers/functions/subscriptions.go
Comment (bolekk) [reply core/services/gateway/handlers/functions/subscriptions.go:146]: We will log errors when that happens right?

File: core/services/gateway/handlers/functions/handler.functions.go
Comment (pinebit) [reply core/services/gateway/handlers/functions/handler.functions.go:137]: changed to what we use in other cases like this: assets.Link.

File: core/scripts/functions/templates/oracle.toml
Comment (bolekk) [inline core/scripts/functions/templates/oracle.toml:38]: outdated

File: core/services/functions/connector_handler.go
Comment (bolekk) [inline core/services/functions/connector_handler.go:32]: Raname to minimumSubscriptionBalance for consistency.

**Review by bolekk — APPROVED** [review]

**Pull request merged**
